### ISSUE TITLE
Test and Document Custom Builds

### ISSUE THREAD

**Issue by clepore** [infor-design/enterprise#417]
**Original Issue:**  https://jira.infor.com/browse/SOHO-7579

We need to provide a system for slimming down IDS builds to only include the components an app developer wishes to implement.  Given a list of component names (fx: "toolbar, button, datepicker, form, validation"), it should be possible to generate a "custom" version of `sohoxi.js` with just the required components and their dependencies. 

- [x] Document the current process for making a custom build.
- [x] Clean up any imports to make it clean and simple.
- [x] Create a CLI-based tool for generating custom builds.
- [x] Explore custom builds for SASS files, linked to the Rollup bundler.
- [x] Should have test coverage for all individual dependencies.
- [x] Deprecate all mapper, and builder dependencies after review with stakeholders.
- [x] Should include tests. Specifically, a test build methods, and if possible the actual build, and the quality of it.

A good testing for this process should do the following:
- Create a custom build of either a single component or multiple components
- Segment out just the relevant functional/e2e tests for only those components and their dependencies
- generate a coverage report for the custom build

**References/Issues/Possible Blockers:**
https://github.com/rollup/rollup-plugin-multi-entry
https://github.com/rollup/rollup/issues/2049
https://github.com/rollup/rollup/issues/2092

```
// ================================================= /
// Soho Xi - Rollup Configuration
// ================================================= /
import resolve from 'rollup-plugin-node-resolve';
import babel from 'rollup-plugin-babel';
import json from 'rollup-plugin-json';
import license from 'rollup-plugin-license';
import multiEntry from 'rollup-plugin-multi-entry';

const bundleBanner = require('./build/generate-bundle-banner');
const getComponents = (listComponent) => {
  if (listComponent) {
    return listComponent.split(',');
  }

  return ['components/index.js'];
};
module.exports = {
  input: getComponents(process.env.ROLLUP_COMPONENTS),
  output: {
    file: 'dist/js/sohoxi.js',
    format: 'iife',
    sourcemap: true,
    sourcemapFile: 'dist/js/sohoxi.js.map',
    name: 'Soho',
    globals: {
      jquery: '$',
      d3: 'd3',
      alert: 'alert',
      console: 'console'
    }
  },
  plugins: [
    multiEntry({ exports: true }),
    resolve(),
    json(),
    babel({
      exclude: 'node_modules/**', // only transpile our source code
      plugins: ['external-helpers']
    }),
    license({
      sourceMap: true,
      banner: bundleBanner
    })
  ],
};
```

This seems to work, and shouldn't be breaking, again test are required for this modification.

```
env ROLLUP_COMPONENTS='components/dropdown/dropdown.js' npx grunt
env KARMA_SPECS='components/dropdown/unit/dropdown-api.spec.js' npm run ci:unit
```

**Comment by jake-harrington** [comment]
The LSF team has been using custom builds in their various forms for years. We manually managed script imports from source before the mapper tool was created, and since the ES6 migration we've had to make a more custom, yet fragile, transition.

We use custom builds for every html page in our application. Some simple pages only need buttons and toolbars so including the entirety of larger controls like the datagrid or tree isn't needed.

Our custom tooling for our project takes these steps:

1. Each html file in our project has a comment like `<!--control-bundle:button,toolbar,dropdown-->` (all names are lowercase), my tooling parses this to get an array of needed controls for that page
1. We run some regular expressions on `src/components/component.js` and `src/components/components.jquery.js` to get the code for exports and imports and make note of the component names
1. We try and match the parsed component names from the two components files to our lowercased names from our html comments to find the matching import/export statements.
1. Thos matching import/export statements of the required controls from step 1 would be concatenated and we overwrite the two component js files with _only those imports and exports_ we need.
1. `grunt build:js`
1. `dist/js/sohoxi.js` is copied to our project's directory with a name based on the requiring html file's path and file name.

To give an example if I have `<!--controls-bundle:button,toolbar,validation-->` in an html file, my tooling will overwrite `components.js` with:

```
export { Button } from './button/button';
export { Toolbar } from './toolbar/toolbar';
export { Validator} from './validation/validator';
```

And `components.jquery.js` with:

```
import './button/button.jquery';
import './toolbar/toolbar.jquery';
import './validation/validation.jquery';
```

This has worked for us for 4.4-4.5 and now trying it out with 4.9 it seems to be working with some changes, but it's extremely fragile and not easily maintainable. My biggest concerns are:

1. Parsing and generating javascript code to run a build is way more complex than I think it needs to be. Any changes or additions to these files that aren't just imports or exports won't be included unless I write extra code in my tooling.
1. Some components only exist in one file and not the other. I need to write case-by-case overrides for when it shouldn't expect something to match in both (I can't just silently ignore non-matches because sometimes component names get changed and it wouldn't properly import it anymore). Sometime since 4.5 now toolbarsearchfield is only in `components.jquery.js` but not in `components.js` for example.
1. Some names are different between the two files and I need case-by-case overrides for that. For validation messages I need to include the export for 'Validator' in `components.js` but 'validation' in `components.jquery.js`.

In short, we just need an easier way, preferably CLI, where we can pass all of the controls we require as-is and a build is generated without having to touch either of the components files. Also if the component names stay the same so we don't have to adjust the component name overrides or our html comments that would be nice.

If you want a demo of our current process let me know!

**Comment by EdwardCoyle** [comment]
Most of the CLI application for this will be coming in a PR for `4.11.0`, but there are some tasks for it that will need to be pushed to another sprint:

- Better targeting of bundled items via the CLI (see #850) 
- TBD

**Comment by CindyMercadoReyes** [comment]
This issue is now resolved. Closing out this GitHub ticket.
